tests/ga/test_env.py

```diff
@@ -0,0 +1,161 @@
+# Copyright 2018 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+from azurelinuxagent.common.osutil.default import DefaultOSUtil, shellutil
+from azurelinuxagent.ga.env import EnvHandler
+from tests.tools import *
+
+
+class TestEnvHandler(AgentTestCase):
+    def setUp(self):
+        AgentTestCase.setUp(self)
+
+        # save the original run_command so that mocks can reference it
+        self.shellutil_run_command = shellutil.run_command
+
+        # save an instance of the original DefaultOSUtil so that mocks can reference it
+        self.default_osutil = DefaultOSUtil()
+
+        # AgentTestCase.setUp mocks osutil.factory._get_osutil; we override that mock for this class with a new mock
+        # that always returns the default implementation.
+        self.mock_get_osutil = patch("azurelinuxagent.common.osutil.factory._get_osutil", return_value=DefaultOSUtil())
+        self.mock_get_osutil.start()
+
+
+    def tearDown(self):
+        self.mock_get_osutil.stop()
+        AgentTestCase.tearDown(self)
+
+    def test_get_dhcp_client_pid_should_return_a_sorted_list_of_pids(self):
+        with patch("azurelinuxagent.common.utils.shellutil.run_command", return_value="11 9 5 22 4 6"):
+            pids = EnvHandler().get_dhcp_client_pid()
+            self.assertEquals(pids, ["11", "22", "4", "5", "6", "9"])
```

Comment:
Just curious, in an ideal case shouldn't the sorting be on the numeric value rather than the string value?

Reply:
Good point. I chatted with Paula about this, but not with you, sorry. There are too many occurrences of run_get_output and I need to look at each one of them to analyze how they should behave. In this PR I did some incremental improvements to run_command and undid the periodic logging in run_get_output (it may make debugging harder). The periodic logging in run_get_output was meant to fix the "pidof" message to the serial console, so I focused on fixing that issue. There are still multiple instances of get_dhcp_client_pid that use run_get_output (I changed only those occurrences that call "pidof") and those return strings so my new implementations also return strings. I only need an order to guarantee the comparison will succeed, but I do not care about whether it is numeric or alphabetic. As I keep replacing more occurrences of run_get_output I will make get_dhcp_client_pid return integers instead of strings and then the order will be numeric.

Reply:
Fair enough, sounds good. Thanks! :)